### PR TITLE
[SEMVER-MAJOR] Replace continuation-local-storage with cls-hooked

### PR DESCRIPTION
**Pull request by josieusa**
Hello, this is a quick and (not so) dirty alternative to PR #2.

Rationale:
- Personally, I still like more the approach I used in PR #2, but I just realized it needs further analysis, because I initially forgot to take into account issue [async-listener#57](https://github.com/othiym23/async-listener/issues/57).
- The quickest and most obvious way to address this is, of course, to get rid of `async-listener`, which already was the long-term goal of my previous PR, anyway. Here it just turned into an immediate goal, that's all.
- One trivial way to reach this goal immediately is to completely replace `continuation-local-storage` with `cls-hooked` **quitting any attempt to keep backward compatibility with the former**, which is exactly what I did in this PR and is the difference with my previous PR. This should make the commit `df60f13`no longer necessary, so I reverted it.
- `cls-hooked`, however (maybe `continuation-local-storage` too, I don't care), needs to be required before everything else, in order to not fail silently and catastrophically. So, I also gave proper and detailed instructions about how to do this in the README.md

Anyway, given the very short time I dedicated to this PR, I would consider this a **work-in-progress** for now.

TODO:
- Bump version to 2.0.0-alpha.1 when it's more ready for alpha release.

PS
For now, in order to make tests work, after deleting `node_modules` folder and running `npm i`, I had to run (NPM v3):

```
npm r strong-remoting
npm i strong-remoting@3.0.0-alpha.5
```

I suppose this is because [loopback PR #2696](https://github.com/strongloop/loopback/pull/2696/commits/6e1defcb1805176867fdf0101e9d5ae64a80ba75) maybe hasn't landed in `loopback-context` yet.

Also, for the same reason, if you test this branch from inside a Loopback app, make sure that `loopback-context` uses the same version of `strong-remoting` as the app.

### CODE REVIEW

**Comment by slnode** [comment]
Can one of the admins verify this patch? To accept patch and trigger a build add comment "._ok\W+to\W+test._"

**Comment by slnode** [comment]
Can one of the admins verify this patch?

File: CHANGES.md [hunk 1]
```diff
@@ -2,3 +2,8 @@
 =========================
 
  * First release!
+
+2016-08-10, Version 2.0.0-alpha.1
```
Comment (bajtos) [inline CHANGES.md]: The contents in `CHANGES.md` is generated by our release tooling, please revert this change.

File: .travis.yml [hunk 2]
```diff
@@ -1,7 +1,4 @@
 sudo: false
 language: node_js
 node_js:
-  - "0.10"
-  - "0.12"
-  - "4"
```
Comment (bajtos) [inline .travis.yml:6]: Whoa, dropping support for 4.x, which is still LTS, and requiring only 6.x, which is not LTS yet. I don't think users will like this.

It's not necessarily a blocker, but we must come up with a reasonable way how to support users running on pre-v6.x versions. Perhaps publish the new version under the `next` tag, so that `npm install loopback-context` keep installing the current `stable` version line `1.x`?

File: package.json [hunk 3]
```diff
@@ -14,15 +14,16 @@
   "main": "server/current-context.js",
   "browser": "browser/current-context.js",
   "scripts": {
-    "test": "mocha",
+    "test": "mocha -r cls-hooked",
```
Comment (bajtos) [inline package.json]: It's better to specify options like this in `test/mocha.opts`. That way it still possible to run the tests via `mocha`. I do that frequently, e.g. when I want to run only a subset of tests.

File: test/main.test.js [hunk 4]
```diff
@@ -5,6 +5,8 @@
 
 'use strict';
 
+// ASYNC VERSION MATTERS! 1.5.2 is required in order for this test to work.
+var async = require('async');
```
Comment (bajtos) [inline test/main.test.js:8]: Please add a version check.

``` js
assert.equals(require('async/package.json').version, '1.5.2');
```

File: .travis.yml (hunk 2 above)
Comment (josieusa) [reply .travis.yml:6]: There is no real need to drop support for 4.x, provided that the user has at least v4.5. I did this because Travis supports only v4.4, according to this link
[Available versions](https://docs.travis-ci.com/user/languages/javascript-with-nodejs/#Available-Versions)
And that would cause the build to fail, I think, because async-hook requires Node >= 4.5.
I don't have Travis and I can't check that v4.5 is really unsupported by Travis, and that Travis docs are not outdated.
I could just try and re-enable v4 and see if the build passes...

Comment (josieusa) [reply .travis.yml:6]: Ok, nevermind, I checked if Travis docs were outdated, then I reverted the change in `.travis.yml` and the build succeeded; however, users must be aware that this branch fails with Node < 4.5.x (I tested it with 4.4.x and it fails).

**Comment by josieusa** [comment]
Ok, I addressed all the line notes from the mantainer; but before removing the [WIP] in the title I'd like to try this in production, despite having some successful tests already in place. I'll let you know soon.

**Comment by bajtos** [comment]
> Ok, I addressed all the line notes from the mantainer; but before removing the [WIP] in the title I'd like to try this in production, despite having some successful tests already in place. I'll let you know soon.

Sounds good, thank you!

**Comment by evseevnn-zz** [comment]
How soon this come?
I tested code from `josieusa:feature/cls-hooked-3`
**This not work!**
I did it like writen in readme.md

Maybe you can prepare worked example or change readme.md?

**UPD. Sorry guys! This work, but problem different, just i use `loopback-connector-mongodb@1.15.2` module. Explaination [here](https://github.com/strongloop/loopback/issues/2519#issuecomment-237173349)**

**Comment by bajtos** [comment]
FWIW, #14 has dropped support for Node v0.10 and Node v0.12. There are no blockers now that would prevent us from landing an implementation based on cls-hooked.

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by josieusa** [comment]
@bajtos Sorry for the late response! Luckily, I'm testing this in production right now, and I'll tell you if it works for me ASAP.
I'd only mention that in production I'm using Node 6 and loopback 2, but if it works on 2, I don't see any reason why it shouldn't work on LB 3, since the legacy context feature is deprecated on both versions (anyway, Mocha tests are run on LB 3 since it's a dev dependency).
Just one caveat: if the app uses Node 4, other stuff wouldn't break because of this fix, but if the app uses the current context API, the legacy context feature would be used as a fallback, which is not (and likely will never be) fixed by this PR.
TL;DR whoever wishes to use the working current context feature in this PR, should update to Node 6 before.
Thank you.

**Comment by bajtos** [comment]
> TL;DR whoever wishes to use the working current context feature in this PR, should update to Node 6 before.

In that case, let's bump up `engines.node` field in `package.json` to `>=6` (or whatever minimal Node.js version is required for your new code to work correctly). If the fallback mechanism is never used on Node v6 (and newer), then I am fine with removing this fallback mechanism completely and perhaps throwing an error if the required Node.js API is not available.

**Comment by josieusa** [comment]
Hello,
I updated the `engines.node` field to `^4.5 || ^5.10 || ^6.0` instead of just 6.0 because this is the exact version range required by `async-hook` dependency of this fork (from its `package.json`).
My tests in production with Node 6 and Loopback 2 are going fine, and they will end around Tuesday.
After the end of the tests, I would like to use this fork permanently in production.
So, since there are no more line notes left to address, what do you think about a merge when my tests end successfully?
I don't need to add more commits, maybe just some edits to `README` before Tuesday.
Thank you.

File: README.md [hunk 5]
```diff
@@ -1,7 +1,44 @@
...
+
+**Do NOT test this package using `mocha .`.**
+
+Test using:
+
+```
+npm test
+```
+
+This adds the `-r` option to `mocha` command, needed in order to pass tests.
```
Comment (bajtos) [inline README.md]: Please move the `-r` option to `test/mocha.opts` file, so that people can run tests using `mocha` directly. I personally use this frequently, e.g. to run only a single test via `mocha -g "test name"`.

```
--require cls-hooked
```

(BTW you pull request does not change `package.json` despite what README says.)

File: package.json [hunk 6]
```diff
@@ -1,9 +1,9 @@
 {
   "name": "loopback-context",
-  "version": "3.0.0-alpha.1",
```
Comment (bajtos) [inline package.json:3]: Please don't change the version number, we will bump it when releasing a new version.

File: test/main.test.js [hunk 7]
```diff
@@ -98,4 +99,40 @@ describe('LoopBack Context', function() {
...
+        var ctx = LoopBackContext.getCurrentContext();
+        expect(ctx).is.an('object');
+        var testValue = ctx && ctx.get('test-key', 'test-value');
+        cb(null, testValue);
+      };
+      // Trigger async waterfall callbacks
+      var asyncFn = function() {
+        async.waterfall([
+          fn1,
+          fn2,
```
Comment (bajtos) [inline test/main.test.js]: FWIW, we usually structure our code differently:

```js
async.waterfall([
  function pushToContext(next) {
    // code from fn1
  },
  function pullFromContext(next) {
    // code from fn2
  },
  function verify(next) {
    expect(testValue).to.equal('test-value');
    next();
  },
], done);
```

File: README.md (hunk 5 above)
Comment (bajtos) [reply README.md]: Ah, I see that you already have `test/mocha.opts`. I think we should simply remove this section from README then.

**Review by bajtos — CHANGES_REQUESTED** [review]
PTAL 👇

**Comment by josieusa** [comment]
Hello,
I confirm it works for me using both usage modes, that is, both `node -r cls-hooked .` and `require('cls-hooked')`. Issues like strongloop/loopback#2397, strongloop/loopback#2519, strongloop/loopback#2485 didn't occur to me yet since using this fork (**UPDATE**: except if using `when` package, see [this issue](https://github.com/strongloop/loopback/issues/2871#issuecomment-269451238)). ~Though the LB issue 2485 should be load-tested.~ Someone did a load test [here](https://github.com/strongloop/loopback/issues/2871#issuecomment-269585814), resulting in improvements but no solution to the `when` issue, though he hasn't shared test code yet. Because of this, I'll keep using the fork in production with `shrinkwrap`, in an average-size app with hundreds of users, regardless of whether it gets merged or not.
Thank you

**Comment by bajtos** [comment]
@josieusa awesome, thank you for the update! I am not sure if I'll have time to review and land this before I leave for vacation.

**Comment by dehypnosis** [comment]
lose a lot of time again. sad.
i have no idea that how many people using loopback deals with current context problems..

**Comment by josieusa** [comment]
@dehypnosis If you want to use this in your app before it gets merged, you can do like this in `package.json` if your NPM version is recent enough:
```
...
"dependencies": {
...
    "loopback-context": "github:josieusa/loopback-context#feature/cls-hooked-3",
...
},
...

```
If you do this, I strongly recommend to edit `package.json` again afterwards, as soon as this PR gets merged.
Please remember that this PR is not considered to be the official solution (though it works fine for me).

**Review by bajtos — APPROVED** [review]
The patch looks good to me now. In preparation for landing it, I have cleaned up the commit history and `git push -f` to your feature branch.

**Comment by bajtos** [comment]
Landed, thank you for your contribution 🎉  and for keeping the ball rolling despite the long time it took to get this work done 🙇

**Comment by michaelfreund** [comment]
@bajtos i think the version number is confusing since 3.0.0 seems to be related to LB3. I guess this is wrong. After some tests it seems that LB2.x projects should also include loopback-context@3.0.0 to get this changes.

**Comment by bajtos** [comment]
> After some tests it seems that LB2.x projects should also include loopback-context@3.0.0 to get this changes.

Yes, you should be able to use `loopback-context@3.0.0` in LB2.x projects too.

> i think the version number is confusing since 3.0.0 seems to be related to LB3. I guess this is wrong. 

I am afraid it would be similarly confusing if we used 2.0.0, because it would hint relation to LB2.

We can improve the documentation though to make this more clear. Can you contribute this change yourself please? @michaelfreund

**Comment by michaelfreund** [comment]
@bajtos I totally agree with the version number. Of course i can contribute. What do you think of including related component information in the docs? Like for the context stuff

**Loopback 3.x**
Latest Component Version: *3.0.0*
**Loopback 2.x**
Latest Component Version: *3.0.0*
**Source**
https://github.com/strongloop/loopback-context

BTW http://loopback.io/doc/en/lb3/Using-current-context.html#use-a-custom-strong-remoting-phase seems also working in LB2.x but is missing in the docs. Can you confirm? So i can update that too.

**Comment by bajtos** [comment]
@michaelfreund sorry for responding late.

> What do you think of including related component information in the docs? Like for the context stuff

I am proposing to add a new section to loopback-context's README with a table similar to
https://github.com/strongloop/loopback-component-explorer#supported-versions

> BTW http://loopback.io/doc/en/lb3/Using-current-context.html#use-a-custom-strong-remoting-phase seems also working in LB2.x but is missing in the docs. Can you confirm? So i can update that too.

Yes, custom remoting phases are available in LB 2.x too.
